### PR TITLE
add mongodb sharding instance support

### PR DESCRIPTION
**Pull request by AJenpan**
add MongoDB sharding instance support
the test result:
=== RUN   TestAccAlicloudMongoDBShardingInstance_import
--- PASS: TestAccAlicloudMongoDBShardingInstance_import (781.59s)
=== RUN   TestAccAlicloudMongoDBShardingInstance_classic
--- PASS: TestAccAlicloudMongoDBShardingInstance_classic (1081.40s)
=== RUN   TestAccAlicloudMongoDBShardingInstance_vpc
--- PASS: TestAccAlicloudMongoDBShardingInstance_vpc (1115.70s)
=== RUN   TestAccAlicloudMongoDBShardingInstance_multi_instance
--- PASS: TestAccAlicloudMongoDBShardingInstance_multi_instance (1122.78s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     4101.500s

### CODE REVIEW

File: alicloud/resource_alicloud_mongodb_sharding_instance.go [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"vswitch_id": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Optional: true,
+				Computed: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance.go]: Why `vswitch_id` is computed?

File: alicloud/resource_alicloud_mongodb_sharding_instance.go [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateDBInstanceName,
+			},
+			"security_ip_list": {
+				Type:     schema.TypeSet,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Computed: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance.go]: Why `security_ip_list ` is computed?

File: alicloud/resource_alicloud_mongodb_sharding_instance.go [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+			"period": {
+				Type:             schema.TypeInt,
+				ValidateFunc:     validateAllowedIntValue([]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 12, 24, 36}),
+				Optional:         true,
+				Computed:         true,
+				DiffSuppressFunc: mongoDBPeriodDiffSuppressFunc,
+			},
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance.go]: I think if the `vswitch_id` is set, the `zone_id` can be optional.

File: alicloud/resource_alicloud_mongodb_sharding_instance.go [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+			return nil, WrapError(Error("The specified vswitch " + vsw.VSwitchId + " isn't in the zone " + request.ZoneId))
+		}
+		request.VSwitchId = vswitchId
+		request.NetworkType = strings.ToUpper(string(Vpc))
+		request.VpcId = vsw.VpcId
+	}
+
+	request.ChargeType = d.Get("instance_charge_type").(string)
+	period, ok := d.GetOk("period")
+	if PayType(request.ChargeType) == PrePaid && ok {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance.go]: I think line 198 should be `if ok && PayType(request.ChargeType) == PrePaid`

File: alicloud/resource_alicloud_mongodb_sharding_instance.go [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+			return WrapError(err)
+		}
+		d.SetPartial("mongo_list")
+	}
+
+	if d.IsNewResource() {
+		d.Partial(false)
+		return resourceAlicloudMongoDBShardingInstanceRead(d, meta)
+	}
+
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance.go]: if `create` return `read` directly, line 322-325 can be removed.

File: alicloud/resource_alicloud_mongodb_sharding_instance_test.go [hunk 6]
```diff
@@ -0,0 +1,950 @@
...
+		id := v.DBInstanceId
+		skip := true
+		for _, prefix := range prefixes {
+			if strings.HasPrefix(strings.ToLower(name), strings.ToLower(prefix)) {
+				skip = false
+				break
+			}
+		}
+		if skip {
+			continue
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance_test.go]: The line 77 should print the skip details

File: alicloud/resource_alicloud_mongodb_sharding_instance_test.go [hunk 7]
```diff
@@ -0,0 +1,950 @@
...
+		}
+
+		sweeped = true
+
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance_test.go]: Please add a print message before delete the instance

File: alicloud/resource_alicloud_vswitch_test.go [hunk 8]
```diff
@@ -35,6 +35,7 @@ func init() {
 			"alicloud_elasticsearch_instance",
 			"alicloud_vpn_gateway",
 			"alicloud_mongodb_instance",
+			"alicloud_mongodb_sharding_instance",
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_vswitch_test.go]: good

File: alicloud/resource_alicloud_mongodb_sharding_instance_test.go [hunk 9]
```diff
@@ -0,0 +1,950 @@
...
+    node_class = "dds.mongos.mid"
+    }, {
+    node_class = "dds.mongos.mid"
+  }]
+  name             = "tf-testAccMongoDBShardingInstance_test_together"
+  account_password = "1234@test"
+  security_ip_list = ["10.168.1.12", "10.168.1.13"]
+}`
+
+func TestAccAlicloudMongoDBShardingInstance_multi_instance(t *testing.T) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance_test.go]: good

File: alicloud/resource_alicloud_mongodb_sharding_instance_test.go [hunk 10]
```diff
@@ -0,0 +1,950 @@
...
+const testMongoDBShardingInstance_multi_instance_together = `
+data "alicloud_zones" "default" {
+  available_resource_creation = "MongoDB"
+}
+variable "name" {
+  default = "tf-testAccMongoDBShardingInstance_multi_instance"
+}
+
+resource "alicloud_mongodb_sharding_instance" "default" {
+  count          = 2
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_mongodb_sharding_instance_test.go]: Please test at least 5 instances.

File: alicloud/service_alicloud_mongodb.go [hunk 11]
```diff
@@ -134,3 +135,106 @@ func (s *MongoDBService) ModifyMongoDBSecurityIps(instanceId, ips string) error
...
+
+			addDebug(request.GetActionName(), raw)
+
+			err = server.WaitForMongoDBInstance(instanceID, Running, DefaultLongTimeout)
+			if err != nil {
+				return WrapError(err)
+			}
+		}
+	} else {
+
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_mongodb.go]: If there is no any code, please remove else

File: website/docs/r/mongodb_sharding_instance.html.markdown [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+Provides a MongoDB sharding instance resource supports replica set instances only. the MongoDB provides stable, reliable, and automatic scalable database services. 
+It offers a full range of database solutions, such as disaster recovery, backup, recovery, monitoring, and alarms.
+You can see detail product introduction [here](https://www.alibabacloud.com/help/doc-detail/26558.htm)
+
+-> **NOTE:**  The following regions don't support create Classic network MongoDB sharding instance.
+[`cn-zhangjiakou`,`cn-huhehaote`,`ap-southeast-2`,`ap-southeast-3`,`ap-southeast-5`,`ap-south-1`,`me-east-1`,`ap-northeast-1`,`eu-west-1`] 
+
+-> **NOTE:**  Create MongoDB Sharding instance or change instance type and storage would cost 10~20 minutes. Please make full preparation
+
```
Comment (xiaozhu36) [inline website/docs/r/mongodb_sharding_instance.html.markdown]: Please add a note "Available in 1.40.0"

File: website/docs/r/mongodb_sharding_instance.html.markdown [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+resource "alicloud_mongodb_sharding_instance" "foo" {  
+  zone_id          = "${data.alicloud_zones.default.zones.0.id}"
+  engine_version   = "3.4"
+  name             = "${var.name}"  
+  shard_list       = ["${var.shard}", "${var.shard}"]
+  mongo_list       = ["${var.mongo}", "${var.mongo}"]
+}
+
```
Comment (xiaozhu36) [inline website/docs/r/mongodb_sharding_instance.html.markdown]: The example had better use VPC network.

File: website/docs/r/mongodb_sharding_instance.html.markdown [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        - Custom storage space; value range: [10, 1,000]
+        - 10-GB increments. Unit: GB. 
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The ID of the MongoDB.
+* `zone_id` - The virtual switch ID 
+* `vswitch_id`  - The virtual switch ID
```
Comment (xiaozhu36) [inline website/docs/r/mongodb_sharding_instance.html.markdown]: i think `zone_id` and `vswitch_id` can be remvoed.

**Comment by AJenpan** [comment]
=== RUN   TestAccAlicloudMongoDBShardingInstance_import
--- PASS: TestAccAlicloudMongoDBShardingInstance_import (752.93s)
=== RUN   TestAccAlicloudMongoDBShardingInstance_classic
--- PASS: TestAccAlicloudMongoDBShardingInstance_classic (1073.09s)
=== RUN   TestAccAlicloudMongoDBShardingInstance_vpc
--- PASS: TestAccAlicloudMongoDBShardingInstance_vpc (1102.48s)
=== RUN   TestAccAlicloudMongoDBShardingInstance_multi_instance
--- PASS: TestAccAlicloudMongoDBShardingInstance_multi_instance (1084.55s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     4013.084s

**Pull request merged**